Mathlib/Tactic/CasesM.lean

```diff
@@ -40,17 +40,24 @@ partial def casesMatching (matcher : Expr → MetaM Bool) (recursive := false) (
             g.cases ldecl.fvarId
           else
             let s ← saveState
-            let subgoals ← g.cases ldecl.fvarId
+            let subgoals ← g.cases ldecl.fvarId (givenNames := #[⟨true, [ldecl.userName]⟩])
             if subgoals.size > 1 then
               s.restore
               continue
             else
               pure subgoals
           for subgoal in subgoals do
+            -- If only one new hypothesis is generated, rename it to the original name.
+            let g ← if h : subgoal.fields.size = 1 then
+              match subgoal.fields[0] with
+              | .fvar fvarId => subgoal.mvarId.rename fvarId ldecl.userName
+              | _ => pure subgoal.mvarId
+            else
+              pure subgoal.mvarId
```

Comment:
```suggestion
            let g ← match subgoal.fields with
              | #[.fvar fvarId] => subgoal.mvarId.rename fvarId ldecl.userName
              | _ => pure subgoal.mvarId
```
maybe?

Reply:
Thanks, very nice, I modified this too many times while writing it.